File: package/umt_python/tests/unit/date/test_format_relative.py
import unittest
from datetime import datetime, timedelta

from src.date import format_relative


class TestFormatRelative(unittest.TestCase):
    def setUp(self):
        self.base = datetime(2025, 4, 15, 12, 0, 0)

    def test_returns_now_for_delta_below_one_second(self):
        target = self.base + timedelta(milliseconds=500)
        self.assertEqual(format_relative(target, self.base, "en"), "now")

    def test_formats_seconds_in_the_past(self):
        target = self.base - timedelta(seconds=5)
        self.assertEqual(format_relative(target, self.base, "en"), "5 seconds ago")

    def test_formats_minutes_in_the_future(self):
        target = self.base + timedelta(minutes=5)
        self.assertEqual(format_relative(target, self.base, "en"), "in 5 minutes")

    def test_formats_hours(self):
        target = self.base - timedelta(hours=3)
        self.assertEqual(format_relative(target, self.base, "en"), "3 hours ago")

    def test_formats_days(self):
        target = self.base + timedelta(days=1)
        self.assertEqual(format_relative(target, self.base, "en"), "tomorrow")

    def test_formats_weeks(self):
        target = self.base - timedelta(days=14)
        self.assertEqual(format_relative(target, self.base, "en"), "2 weeks ago")

    def test_formats_months(self):
        target = self.base + timedelta(days=90)
        self.assertEqual(format_relative(target, self.base, "en"), "in 3 months")

    def test_formats_years(self):
        target = self.base - timedelta(days=2 * 365)
        self.assertEqual(format_relative(target, self.base, "en"), "2 years ago")

    def test_uses_default_base_date_when_omitted(self):
        target = datetime.now() + timedelta(seconds=5)
        result = format_relative(target, None, "en")
        self.assertIsInstance(result, str)
        self.assertGreater(len(result), 0)

    def test_formats_japanese_locale(self):
        self.assertEqual(
            format_relative(self.base + timedelta(days=1), self.base, "ja"), "明日"
        )
        self.assertEqual(
            format_relative(self.base - timedelta(seconds=5), self.base, "ja"),
            "5 秒前",
        )
        self.assertEqual(
            format_relative(self.base + timedelta(days=90), self.base, "ja"),
            "3 か月後",
        )


if __name__ == "__main__":
    unittest.main()

File: package/umt_rust/tests/date/test_format_relative.rs
//! Tests for the format_relative module.

use chrono::{Duration, TimeZone, Utc};
use umt_rust::date::umt_format_relative;

fn base() -> chrono::DateTime<Utc> {
    Utc.with_ymd_and_hms(2025, 4, 15, 12, 0, 0).unwrap()
}

#[test]
fn test_returns_now_for_delta_below_one_second() {
    let base = base();
    let target = base + Duration::milliseconds(500);
    assert_eq!(umt_format_relative(&target, &base, Some("en")), "now");
}

#[test]
fn test_formats_seconds_in_the_past() {
    let base = base();
    let target = base - Duration::seconds(5);
    assert_eq!(
        umt_format_relative(&target, &base, Some("en")),
        "5 seconds ago"
    );
}

#[test]
fn test_formats_minutes_in_the_future() {
    let base = base();
    let target = base + Duration::minutes(5);
    assert_eq!(
        umt_format_relative(&target, &base, Some("en")),
        "in 5 minutes"
    );
}

#[test]
fn test_formats_hours() {
    let base = base();
    let target = base - Duration::hours(3);
    assert_eq!(
        umt_format_relative(&target, &base, Some("en")),
        "3 hours ago"
    );
}

#[test]
fn test_formats_days() {
    let base = base();
    let target = base + Duration::days(1);
    assert_eq!(umt_format_relative(&target, &base, Some("en")), "tomorrow");
}

#[test]
fn test_formats_weeks() {
    let base = base();
    let target = base - Duration::days(14);
    assert_eq!(
        umt_format_relative(&target, &base, Some("en")),
        "2 weeks ago"
    );
}

#[test]
fn test_formats_months() {
    let base = base();
    let target = base + Duration::days(90);
    assert_eq!(
        umt_format_relative(&target, &base, Some("en")),
        "in 3 months"
    );
}

#[test]
fn test_formats_years() {
    let base = base();
    let target = base - Duration::days(2 * 365);
    assert_eq!(
        umt_format_relative(&target, &base, Some("en")),
        "2 years ago"
    );
}

#[test]
fn test_default_base_date_when_omitted() {
    let now = Utc::now();
    let target = now + Duration::seconds(5);
    let result = umt_format_relative(&target, &now, Some("en"));
    assert!(!result.is_empty());
}
